### PR TITLE
Recursive function

### PR DESCRIPTION
**Pull request by frenchfrog42**

### CODE REVIEW

File: contracts/final_loop_import.scrypt [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+//DO NOT MODIFY
+struct DoNotTouch {
+    Object computation;
+    bool recursion;
+}
+
+library Aux {
+    static const int fee = 5000;
+    static const int SizeOfObject = 2;
+    static function my_great_function(Object object): DoNotTouch {
```
Comment (xhliu) [inline contracts/final_loop_import.scrypt:30]: Add an empty line

File: contracts/final_loop_import.scrypt [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+
+//Please modify
+struct Object {
+    int a;
+    int b;
+}
+
+//DO NOT MODIFY
+struct DoNotTouch {
```
Comment (xhliu) [inline contracts/final_loop_import.scrypt:22]: Rename DoNotTouch -> RecursiveReturn

File: contracts/final_loop_import.scrypt [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        if (b == 0) {
+            recursion_call = false;
+        } else {
+            int t = b;
+            b = a % b;
+            a = t;
+        }
+        return {{a, b}, recursion_call};
+    }
+    static function bytes_to_object(bytes b): Object {
```
Comment (xhliu) [inline contracts/final_loop_import.scrypt:42]: rename to deserialized

File: contracts/final_loop_import.scrypt [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        return {unpack(b[0:1]), unpack(b[1:2])};
+    }
+    static function object_to_bytes(Object object): bytes {
```
Comment (xhliu) [inline contracts/final_loop_import.scrypt:45]: serialize

File: contracts/final_loop_import.scrypt [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        return num2bin(object.a, 1) + num2bin(object.b, 1);
+    }
+}
+
+
+/*
```
Comment (xhliu) [inline contracts/final_loop_import.scrypt:51]: delete dead code below

File: contracts/final_loop_main.scrypt [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+//Code of a contract that uses the library
+contract Final {
+    public function unlock_utxo(SigHashPreimage txPreImage) {
+        require(Tx.checkPreimage(txPreImage));
+        Object object = Loop.getState(txPreImage);
+        Object computation = Loop.recursive(txPreImage, object); //This line does a loop
+        //Do what you want after
+        require(true);
+    }
+    //Do not touch below
```
Comment (xhliu) [inline contracts/final_loop_main.scrypt:13]: Add an empty line

File: contracts/final_loop_library.scrypt [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+library Loop {
+	///////////////////////////////////////////////////////////
+	// Interesting part below
+	///////////////////////////////////////////////////////////
+	static function scriptCode(SigHashPreimage preimage): bytes {
+		//Return everything util.scriptCode except the space we need in this library
+		return Util.readVarint(preimage[104:len(preimage)-Aux.SizeOfObject-1]);
+	}
+	static function buildOutput(SigHashPreimage preimage, bytes outputScript, int outputSatoshis): bytes {
```
Comment (xhliu) [inline contracts/final_loop_library.scrypt:13]: Add empty line before each new function.

**Review by xhliu — CHANGES_REQUESTED** [review]
https://github.com/sCrypt-Inc/boilerplate/tree/recurse

Writing is inconsistent with `final_loop_main.scrypt`.

File: contracts/final_loop_library.scrypt (hunk 7 above)
Comment (marcelgruber) [reply contracts/final_loop_library.scrypt:13]: @xhliu would it make sense to automate some of these checks with a linter? Or is there already one in place as part of the build / test process?

Comment (xhliu) [reply contracts/final_loop_library.scrypt:13]: Good idea. Will add one.

**Review by xhliu — APPROVED** [review]
💯 well done. This is huge.

File: contracts/final_loop_library.scrypt [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+		int scriptLen = len(scriptCode);
+		return unpack(scriptCode[scriptLen-1-Aux.SizeOfObject:scriptLen-Aux.SizeOfObject]);
+	}
+
+	static function getTempState(SigHashPreimage preimage): Object {
+        bytes scriptCode = Util.scriptCode(preimage);
+		int scriptLen = len(scriptCode);
+		return Aux.deserialize(scriptCode[scriptLen-Aux.SizeOfObject:]);
+	}
+	//End get state
```
Comment (marcelgruber) [inline contracts/final_loop_library.scrypt:40]: Comment can be removed

**Pull request merged**